docs/implementation-details.md

```diff
@@ -0,0 +1,71 @@
+TODO: Introduce the document
+
+### Structure
+The data structure of the sortition pool is a root node, which is a single
+`uint256`, followed by 6 layers of branch nodes, followed by a leaf layer. Each
+branch node is a `uint256`, and the branch structure is implemented as a
+`mapping(uint256 => mapping(uint256 => uint256))`. The first index represents
+the layer, and the second index represents the position within the layer.
+
+The first branch layer has 8 branches, and every subsequent layer has 8x the number of branches as the previous layer, so 
+
+1) node layer: 1
+2) branch layer 1: 8
+3) branch layer 2: 64
+4) branch layer 3: 512
+5) branch layer 4: 4,096
+6) branch layer 5: 32,768
+7) branch layer 6: 262,144
+8) leaf layer: 2,097,152
+
+### Leaf Serialization and Deserialization
+
+```
+function make(
+  address _operator,
+  uint256 _creationBlock,
+  uint256 _id
+) internal pure returns (uint256) {
+  assert(_creationBlock <= type(uint64).max);
+  assert(_id <= type(uint32).max);
+  uint256 op = uint256(bytes32(bytes20(_operator)));
+  uint256 uid = _id & ID_MAX;
+  uint256 cb = (_creationBlock & BLOCKHEIGHT_MAX) << ID_WIDTH;
+  return (op | cb | uid);
+}
+```
+
+The `Leaf.make` takes in an operator address, the block it was created in, and
+a monotonically increasing id for unique operators. We convert the operator
+address to a bytes20, and then convert it to a bytes32 (which pads the extra 12
+bytes with 0's on the *right*), and then convert the whole thing into a
+`uint256`. This means that we have 160 bits that matter on the left, and 96 0's
+on the right stored in `op`.
+
+`ID_MAX = 2^32 - 1` is represented as 32 1's, which as a `uint256` is 224 0's
+followed by 32 1's. The bitwise `&` clears out everything but the last 32
+significant bits of `_id`, leaving us with 224 0's and 32 significant bits
+stored in `uid`.
```

Comment:
I actually asked the same question when I was going over the code:

```
[4:29 PM] Beau: Given that we require that the operator's id fits within type(uint32).max, I would suspect that it has 224 zeroes in front (at least)
[4:30 PM] Beau: In that light, is there ever a case where
uint256 uid = _id & ID_MAX;

does anything?
```

answer:

> there shouldn't ever be a case where it matters; the code was written with an abundance of caution

Reply:
Oh, I see, I should have paid more attention to the Sortition pool learning Discord thread... 😃 
If both you and me had this question, other readers may be confused as well. It may be worth to mention that this is abundant condition...

Reply:
[ab28a86](https://github.com/keep-network/sortition-pools/pull/164/commits/ab28a86cb5c314a437f3fb10f2628010bb8bee2c)